File: eveil/ios/WordEndCore/Sources/WordEndCore/Policy.swift
// Policy.swift — la pédagogie, séparée de l'acoustique. Portage de
// `eveil/reference/wordend/policy.py` et du garde-fou lexical.
//
// Règles encodées (et testées côté Python comme côté Swift) :
// 1. On ne montre du doigt le fourgon (consonne finale) que sur un verdict
//    `.missingFinalConsonant` — jamais sur `.unsure`.
// 2. Incertain / aucune parole ⇒ messages neutres, jamais « ce n'est pas ça ».
// 3. Syllabe(s) en moins ⇒ aucun wagon désigné : on ne sait pas LAQUELLE
//    manque ; on remodèle le mot entier.
// 4. Pas de boucle d'échec : après `maxAttempts`, on félicite l'effort et on passe.
// 5. Séances courtes, fin ritualisée, budget quotidien réglable par le parent.

import Foundation

public enum CabooseState: String, Sendable {
    case hooked      // accroché : la consonne finale a été entendue
    case pointed     // resté en gare, la mascotte le montre (verdict sûr seulement)
    case waiting     // neutre : en attente
    case none        // mot sans consonne finale
}

public enum MascotAction: String, Sendable {
    case celebrate
    case pointCaboose = "point_caboose"
    case modelWord = "model_word"
    case listenAgain = "listen_again"
    case invite
}

public enum MessageKey: String, CaseIterable, Sendable {
    case bravo
    case hookCaboose = "hook_caboose"
    case modelSlowly = "model_slowly"
    case listenAgain = "listen_again"
    case invite
    case effortNext = "effort_next"

    /// Clés autorisées quand le détecteur n'est PAS sûr : aucune ne désigne d'erreur.
    public static let neutral: Set<MessageKey> = [.listenAgain, .invite, .effortNext, .modelSlowly]

    /// Textes par défaut (l'app utilise un String Catalog ; ceci sert aux tests et aux aperçus).
    public func defaultText(locale: String) -> String {
        let fr: [MessageKey: String] = [
            .bravo: "Bravo ! Tout le train est parti !",
            .hookCaboose: "Oh, le petit fourgon est resté en gare ! On l'accroche avec « chhh » ?",
            .modelSlowly: "Écoute bien le train : on le dit doucement, ensemble.",
            .listenAgain: "Je n'ai pas bien entendu. On le redit ensemble ?",
            .invite: "À toi ! Dis le mot au petit train.",
            .effortNext: "Merci d'avoir bien essayé ! On va voir le mot suivant.",
        ]
        let en: [MessageKey: String] = [
            .bravo: "Hooray! The whole train is leaving!",
            .hookCaboose: "Oh, the little caboose stayed at the station! Let's hook it with \"shhh\"?",
            .modelSlowly: "Listen to the train: let's say it slowly, together.",
            .listenAgain: "I didn't hear it well. Shall we say it again together?",
            .invite: "Your turn! Tell the word to the little train.",
            .effortNext: "Thanks for trying so hard! Let's see the next word.",
        ]
        return (locale.hasPrefix("fr") ? fr : en)[self]!
    }
}

public struct Feedback: Equatable, Sendable {
    public let litWagons: [Bool]
    public let caboose: CabooseState
    public let mascot: MascotAction
    public let message: MessageKey
    public let advance: Bool
    public let countsAsAttempt: Bool
}

public enum FeedbackPolicy {
    public static let maxAttempts = 3

    /// `attempt` commence à 1 pour la première tentative sur ce mot.
    public static func feedback(for verdict: Verdict, wagons: Int, hasCaboose: Bool, attempt: Int,
                                maxAttempts: Int = FeedbackPolicy.maxAttempts) -> Feedback {
        let lastTry = attempt >= maxAttempts
        let allLit = [Bool](repeating: true, count: wagons)
        let noneLit = [Bool](repeating: false, count: wagons)
        let waiting: CabooseState = hasCaboose ? .waiting : .none
        switch verdict.kind {
        case .complete:
            return Feedback(litWagons: allLit, caboose: hasCaboose ? .hooked : .none, mascot: .celebrate,
                            message: .bravo, advance: true, countsAsAttempt: true)
        case .missingFinalConsonant where hasCaboose:
            if lastTry {
                return Feedback(litWagons: allLit, caboose: .waiting, mascot: .modelWord,
                                message: .effortNext, advance: true, countsAsAttempt: true)
            }
            return Feedback(litWagons: allLit, caboose: .pointed, mascot: .pointCaboose,
                            message: .hookCaboose, advance: false, countsAsAttempt: true)
        case .fewerSyllables:
            return Feedback(litWagons: noneLit, caboose: waiting, mascot: .modelWord,
                            message: lastTry ? .effortNext : .modelSlowly, advance: lastTry,
                            countsAsAttempt: true)
        case .noSpeech:
            return Feedback(litWagons: noneLit, caboose: waiting, mascot: .invite,
                            message: lastTry ? .effortNext : .invite, advance: lastTry,
                            countsAsAttempt: true)
        default:
            // Incertain (et tout cas imprévu) : neutre, rien de désigné.
            return Feedback(litWagons: noneLit, caboose: waiting, mascot: .listenAgain,
                            message: lastTry ? .effortNext : .listenAgain, advance: lastTry,
                            countsAsAttempt: true)
        }
    }
}

public enum SessionStep: String, Sendable {
    case `continue`
    case endSession = "end_session"   // rituel de fin (le train rentre au dépôt)
    case restToday = "rest_today"     // à demain
}

/// Durées par défaut prudentes (cf. recommandations écrans, docs/EVEIL.md §7).
public struct SessionPolicy: Equatable, Sendable {
    public var wordsPerSession = 6
    public var maxSessionSeconds = 8.0 * 60
    public var dailyBudgetSeconds = 15.0 * 60     // réglable dans l'espace parent
    public init() {}

    public func nextStep(wordsDone: Int, sessionSeconds: Double, todaySeconds: Double) -> SessionStep {
        if todaySeconds >= dailyBudgetSeconds { return .restToday }
        if wordsDone >= wordsPerSession || sessionSeconds >= maxSessionSeconds { return .endSession }
        return .continue
    }
}

// MARK: - Garde-fou lexical (reconnaissance vocale on-device, OPTIONNELLE)

public enum LexicalEvidence: String, CaseIterable, Sendable {
    case notChecked = "not_checked"      // pas d'ASR, ou transcription vide
    case consistent                      // le mot (ou sa forme tronquée) a été reconnu
    case inconsistent                    // un autre mot semble avoir été dit
}

/// L'ASR ne peut que rendre le verdict PLUS prudent, jamais l'inverse : elle
/// « corrige » vers des mots réels et se trompe beaucoup sur les voix de 3-5 ans.
public func applyLexical(_ v: Verdict, _ evidence: LexicalEvidence) -> Verdict {
    let decisive: Set<VerdictKind> = [.complete, .missingFinalConsonant, .fewerSyllables]
    guard evidence == .inconsistent, decisive.contains(v.kind) else { return v }
    var out = v
    out.kind = .unsure
    out.reasons.append("other_word_suspected")
    return out
}

File: eveil/reference/wordend/policy.py
"""Politique de feedback et de séance — la pédagogie, séparée de l'acoustique.

Le détecteur *mesure* ; cette couche *décide quoi montrer* à un enfant de 3 ans.
Règles encodées (et testées) :

1. On ne **montre du doigt** une partie manquante (le fourgon décroché) que sur
   un verdict MISSING_FINAL_CONSONANT — jamais sur INCERTAIN.
2. INCERTAIN / AUCUNE PAROLE ⇒ messages **neutres** (« on le redit ensemble ? »),
   jamais « ce n'est pas ça ».
3. Syllabe(s) manquante(s) ⇒ on **ne désigne pas** de wagon précis : en français,
   l'enfant tronque souvent la syllabe *initiale* (« nane » pour « banane »),
   donc « allumer le premier wagon » serait souvent faux. On **remodèle** le mot
   entier, wagon par wagon.
4. Pas de boucle d'échec : au-delà de `max_attempts`, on félicite l'effort et on
   passe au mot suivant.
5. Séances courtes, fin ritualisée, budget quotidien réglable par le parent.
"""
from __future__ import annotations

from dataclasses import dataclass

from .detector import Verdict, VerdictKind


@dataclass(frozen=True)
class Feedback:
    lit_wagons: tuple[bool, ...]     # wagons (syllabes) allumés
    caboose: str                     # hooked | pointed | waiting | none (mot sans fourgon)
    mascot: str                      # celebrate | point_caboose | model_word | listen_again | invite
    message_key: str                 # clé de texte localisée (FR/EN), cf. MESSAGES
    advance: bool                    # passer au mot suivant
    counts_as_attempt: bool


# Clés autorisées quand le détecteur n'est PAS sûr : aucune ne désigne d'erreur.
NEUTRAL_KEYS = frozenset({"listen_again", "invite", "effort_next", "model_slowly"})
MAX_ATTEMPTS = 3

MESSAGES = {
    "fr-FR": {
        "bravo": "Bravo ! Tout le train est parti !",
        "hook_caboose": "Oh, le petit fourgon est resté en gare ! On l'accroche avec « chhh » ?",
        "model_slowly": "Écoute bien le train : on le dit doucement, ensemble.",
        "listen_again": "Je n'ai pas bien entendu. On le redit ensemble ?",
        "invite": "À toi ! Dis le mot au petit train.",
        "effort_next": "Merci d'avoir bien essayé ! On va voir le mot suivant.",
    },
    "en-US": {
        "bravo": "Hooray! The whole train is leaving!",
        "hook_caboose": "Oh, the little caboose stayed at the station! Let's hook it with \"shhh\"?",
        "model_slowly": "Listen to the train: let's say it slowly, together.",
        "listen_again": "I didn't hear it well. Shall we say it again together?",
        "invite": "Your turn! Tell the word to the little train.",
        "effort_next": "Thanks for trying so hard! Let's see the next word.",
    },
}


def feedback_for(verdict: Verdict, wagons: int, has_caboose: bool, attempt: int,
                 max_attempts: int = MAX_ATTEMPTS) -> Feedback:
    """`attempt` commence à 1 pour la première tentative sur ce mot."""
    last_try = attempt >= max_attempts
    all_lit = tuple([True] * wagons)
    none_lit = tuple([False] * wagons)
    waiting = "waiting" if has_caboose else "none"
    k = verdict.kind

    if k is VerdictKind.COMPLETE:
        return Feedback(all_lit, "hooked" if has_caboose else "none", "celebrate", "bravo",
                        advance=True, counts_as_attempt=True)
    if k is VerdictKind.MISSING_FINAL_CONSONANT and has_caboose:
        if last_try:
            return Feedback(all_lit, "waiting", "model_word", "effort_next",
                            advance=True, counts_as_attempt=True)
        return Feedback(all_lit, "pointed", "point_caboose", "hook_caboose",
                        advance=False, counts_as_attempt=True)
    if k is VerdictKind.FEWER_SYLLABLES:
        return Feedback(none_lit, waiting, "model_word",
                        "effort_next" if last_try else "model_slowly",
                        advance=last_try, counts_as_attempt=True)
    if k is VerdictKind.NO_SPEECH:
        return Feedback(none_lit, waiting, "invite",
                        "effort_next" if last_try else "invite",
                        advance=last_try, counts_as_attempt=True)
    # INCERTAIN (et tout cas imprévu) : neutre, rien de désigné.
    return Feedback(none_lit, waiting, "listen_again",
                    "effort_next" if last_try else "listen_again",
                    advance=last_try, counts_as_attempt=True)


@dataclass(frozen=True)
class SessionPolicy:
    """Durées par défaut prudentes (cf. recommandations écrans, docs/EVEIL.md §7)."""

    words_per_session: int = 6
    max_session_s: int = 8 * 60
    daily_budget_s: int = 15 * 60        # réglable dans l'espace parent

    def next_step(self, words_done: int, session_s: float, today_s: float) -> str:
        """`continue` | `end_session` (rituel de fin) | `rest_today` (à demain)."""
        if today_s >= self.daily_budget_s:
            return "rest_today"
        if words_done >= self.words_per_session or session_s >= self.max_session_s:
            return "end_session"
        return "continue"
